daemon/src/projection.rs

```diff
@@ -226,6 +237,34 @@ impl Aggregated {
 
         Some(extract_payout_amount(tx, script))
     }
+
+    /// Derive Cfd state based on aggregated state from the events and the
+    /// protocol state
+    fn derive_cfd_state(&self, role: Role) -> CfdState {
+        let mut state = self.state;
+
+        if let Some(settlement_state) = self.settlement_state {
+            state = match settlement_state {
+                ProtocolNegotiationState::Started => match role {
+                    Role::Maker => CfdState::IncomingSettlementProposal,
+                    Role::Taker => CfdState::OutgoingSettlementProposal,
+                },
+                ProtocolNegotiationState::Accepted => CfdState::PendingClose,
+            }
+        };
+        // Rollover is applied after collab settlement for the UI, as it's a
+        // time sensitive and short-lived transition
+        if let Some(rollover_state) = self.rollover_state {
+            state = match rollover_state {
+                ProtocolNegotiationState::Started => match role {
+                    Role::Maker => CfdState::IncomingRolloverProposal,
+                    Role::Taker => CfdState::OutgoingRolloverProposal,
+                },
+                ProtocolNegotiationState::Accepted => CfdState::ContractSetup,
+            }
```

Comment:
Could do this first and just back out of the function early?

Reply:
makes sense - I could do that, and if I return from both if lets I don't need to have the `mut` variable either!